server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java

```diff
@@ -0,0 +1,245 @@
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.mailet.filter;
+
+import static org.apache.james.jmap.api.filtering.Rule.Condition;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import javax.mail.Address;
+import javax.mail.Message;
+import javax.mail.internet.AddressException;
+import javax.mail.internet.InternetAddress;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.james.javax.AddressHelper;
+import org.apache.james.jmap.api.filtering.Rule;
+import org.apache.james.jmap.api.filtering.Rule.Condition.Field;
+import org.apache.mailet.Mail;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.functions.ThrowingFunction;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+public interface MailMatcher {
+
+    interface HeaderExtractor extends ThrowingFunction<Mail, Stream<String>> {}
+
+    class HeaderMatcher implements MailMatcher {
+
+        private final Logger logger = LoggerFactory.getLogger(HeaderMatcher.class);
+
+        private final ContentMatcher contentMatcher;
+        private final String ruleValue;
+        private final HeaderExtractor headerExtractor;
+
+        HeaderMatcher(Optional<ContentMatcher> contentMatcherOptional, String ruleValue,
+                      Optional<HeaderExtractor> headerExtractorOptional) {
+            Preconditions.checkArgument(contentMatcherOptional.isPresent());
+            Preconditions.checkArgument(headerExtractorOptional.isPresent());
+
+            this.contentMatcher = contentMatcherOptional.get();
+            this.ruleValue = ruleValue;
+            this.headerExtractor = headerExtractorOptional.get();
+        }
+
+        @Override
+        public boolean match(Mail mail) {
+            try {
+                Stream<String> headerLines = headerExtractor.apply(mail);
+                return contentMatcher.match(headerLines, ruleValue);
+            } catch (Exception e) {
+                logger.error("error while extracting mail header", e);
+                return false;
+            }
+        }
+    }
+
+    interface ContentMatcher {
+
+        class AddressHeader {
+
+            static class Builder {
+                private final Logger logger = LoggerFactory.getLogger(Builder.class);
+
+                private InternetAddress internetAddress;
+
+                private Builder() {
+                }
+
+                private Builder internetAddress(String addressAsString) {
+                    try {
+                        this.internetAddress = new InternetAddress(addressAsString);
+                    } catch (AddressException e) {
+                        logger.error("error while parsing address " + addressAsString, e);
```

Comment:
Use placeholder for your logging....

Reply:
MM I am sceptical about this error handling, as internetAdress will be null and the error will be less explicit, while calling build...

If you don't want to build a broken object, propagating seems a better choice IMO (at least in a first time) or build an `Optional<AddressHeader>` below